crates/matrix-sdk-base/src/store/indexeddb_store.rs

```diff
@@ -0,0 +1,1043 @@
+            }
+            Ok(())
+        }));
+
+        let db: IdbDatabase = db_req.into_future().await?;
+
+        let tx: IdbTransaction =
+            db.transaction_on_one_with_mode("matrix-sdk-state", IdbTransactionMode::Readwrite)?;
+        let ob = tx.object_store("matrix-sdk-state")?;
+
+        let store_key: Option<DatabaseType> = ob
+            .get(&JsValue::from_str(KEYS::STORE_KEY))?
+            .await?
+            .map(|k| k.into_serde())
+            .transpose()?;
+
+        let store_key = if let Some(key) = store_key {
+            if let DatabaseType::Encrypted(k) = key {
+                StoreKey::import(passphrase, k).map_err(|_| StoreError::StoreLocked)?
+            } else {
+                return Err(StoreError::UnencryptedStore);
+            }
+        } else {
+            let key = StoreKey::new().map_err::<StoreError, _>(|e| e.into())?;
+            let encrypted_key = DatabaseType::Encrypted(
+                key.export(passphrase).map_err::<StoreError, _>(|e| e.into())?,
+            );
+            ob.put_key_val(
+                &JsValue::from_str(KEYS::STORE_KEY),
+                &JsValue::from_serde(&encrypted_key)?,
+            )?;
+            key
+        };
+
+        tx.await.into_result()?;
+
+        IndexeddbStore::open_helper(name, Some(store_key)).await
+    }
+
+    pub async fn open_with_name(name: String) -> Result<Self> {
+        IndexeddbStore::open_helper(name, None).await
+    }
+
+    fn serialize_event(&self, event: &impl Serialize) -> Result<JsValue, SerializationError> {
+        Ok(match self.store_key {
+            Some(ref key) => JsValue::from_serde(&key.encrypt(event)?)?,
+            None => JsValue::from_serde(event)?,
+        })
+    }
+
+    fn deserialize_event<T: for<'b> Deserialize<'b>>(
+        &self,
+        event: JsValue,
+    ) -> Result<T, SerializationError> {
+        match self.store_key {
+            Some(ref key) => {
+                let encrypted: EncryptedEvent = event.into_serde()?;
+                Ok(key.decrypt(encrypted)?)
+            }
+            None => Ok(event.into_serde()?),
+        }
+    }
+
+    pub async fn save_filter(&self, filter_name: &str, filter_id: &str) -> Result<()> {
+        let tx = self
+            .inner
+            .transaction_on_one_with_mode(KEYS::SESSION, IdbTransactionMode::Readwrite)?;
+
+        let obj = tx.object_store(KEYS::SESSION)?;
+
+        obj.put_key_val(
+            &JsValue::from_str(&format!("{}:{}", KEYS::FILTER, filter_name)),
+            &JsValue::from_str(filter_id),
+        )?;
+
+        tx.await.into_result()?;
+
+        Ok(())
+    }
+
+    pub async fn get_filter(&self, filter_name: &str) -> Result<Option<String>> {
+        Ok(self
+            .inner
+            .transaction_on_one_with_mode(KEYS::SESSION, IdbTransactionMode::Readonly)?
+            .object_store(KEYS::SESSION)?
+            .get(&JsValue::from_str(&format!("{}:{}", KEYS::FILTER, filter_name)))?
```

Comment:
> Is there a reason the Encode trait that is used in the Sled store was abandoned.

Yes, IndexedDB requires strings as keys and JsValue simply doesn't have any notion of a `Vec<u8>` to convert from. Thus taking the extra step of encoding str -> vec<u8> to then go back to str didn't sound that useful.

> Another thing, the `:` is used as a separator, this is fine as long as we don't decode things or use the colon to indicate a range.

We only ever use the key for lookup, we never split or try to read content from it, we never interpret its parts. Content isn't stored like this and the only keys we ever read and decode are unprefixed `UserId`s. When using ranges we take the entire range from `PREFIX:` to `PREFIX;` for which it is only relevant that `;` comes after `:`, thus including anything that is having that prefix as no other character can come at that position with that prefix.

> Since this hack isn't available in IndexedDB, I think, we'll need to escape our separator.

Under which circumstance do we need to escape the separator? Like even with `:` in the username in the tests, there is no need. I don't see any reason to escape the `:`. If there is any, I gladly add a test for that and make sure we have a robust system.

> In the Sled store we use the fact that the 0xff byte isn't valid in utf-8 so it's safe to use it as a separator.

I guess this is technically still possible to do, but yeah, it requires us to add `unsafe` code:

```
const ENCODE_SEPARATOR: u8 = 0xff;

trait EncodeKey {
    fn encode(&self) -> JsValue;
}

impl EncodeKey for (&str, &str, &str, &str) {
    fn encode(&self) -> JsValue {
        unsafe {
        JsValue::from_str(
            std::str::from_utf8_unchecked(&[
                self.0.as_bytes(),
                &[ENCODE_SEPARATOR],
                self.1.as_bytes(),
                &[ENCODE_SEPARATOR],
                self.2.as_bytes(),
                &[ENCODE_SEPARATOR],
                self.3.as_bytes(),
                &[ENCODE_SEPARATOR],
            ].concat()))

        }
    }
}
```

And considering that we'd only be doing that introduce an unsafe character in into an otherwise safe string, I don't see any benefit in doing those extra loops.

Reply:
> Yes, IndexedDB requires strings as keys and JsValue simply doesn't have any notion of a `Vec<u8>` to convert from. Thus taking the extra step of encoding str -> vec to then go back to str didn't sound that useful.

Oh sorry, I wasn't clear enough there. I didn't mean that we should use the same `Encode` trait, but have our own `Encode` trait that outputs a `String` instead of a `Vec<u8>`. As mentioned, I thing that's a bit more readable and less error prone than the `format!()` calls we have currently.

> I guess this is technically still possible to do, but yeah, it requires us to add `unsafe` code:

I don't think we should use `unsafe` for any of this.